File: backend/app/repositories/logs.py
from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


class LogsRepository:
    def __init__(self, *, supabase: Any):
        self._supa = supabase

    def insert_visit(
        self,
        *,
        ip: str,
        path: str,
        method: str,
        user_agent: str | None,
        referer: str | None,
        request_id: str | None,
    ) -> None:
        payload = {
            "ip": (ip or "unknown")[:100],
            "path": (path or "").strip()[:500],
            "method": (method or "").strip()[:20],
            "user_agent": (user_agent or "").strip()[:500] or None,
            "referer": (referer or "").strip()[:500] or None,
            "request_id": (request_id or "").strip()[:100] or None,
        }

        try:
            self._supa.table("visit_logs").insert(payload).execute()
        except Exception:
            logger.exception("Failed to insert visit log")

    def insert_chat_log(
        self,
        *,
        ip: str,
        request_id: str | None,
        question: str,
        history: list[dict],
        response_text: str | None,
        sources: list[dict] | None,
        query_plan: dict | None,
        model: str | None,
        status: str,
        error_message: str | None = None,
    ) -> None:
        payload = {
            "ip": (ip or "unknown")[:100],
            "request_id": (request_id or "").strip()[:100] or None,
            "question": (question or "").strip()[:20_000],
            "history": history or [],
            "response_text": (response_text or "")[:80_000] or None,
            "sources": sources or [],
            "query_plan": query_plan,
            "model": (model or "").strip()[:100] or None,
            "status": (status or "").strip()[:50] or "unknown",
            "error_message": (error_message or "").strip()[:2_000] or None,
        }

        try:
            self._supa.table("chat_logs").insert(payload).execute()
        except Exception:
            logger.exception("Failed to insert chat log")

File: backend/internal/repo/logs.go
package repo

import (
	"log/slog"

	"yunews/backend/internal/supabase"
)

// LogsRepository handles visit_logs and chat_logs tables.
type LogsRepository struct {
	db *supabase.Client
}

// NewLogsRepository creates a new LogsRepository.
func NewLogsRepository(db *supabase.Client) *LogsRepository {
	return &LogsRepository{db: db}
}

// InsertVisit inserts a visit log row. Best-effort — errors are only logged.
func (r *LogsRepository) InsertVisit(ip, path, method string, userAgent, referer, requestID *string) {
	payload := map[string]interface{}{
		"ip":         truncate(normalizeIP(ip), 100),
		"path":       truncate(path, 500),
		"method":     truncate(method, 20),
		"user_agent": truncateOrNilPtr(userAgent, 500),
		"referer":    truncateOrNilPtr(referer, 500),
		"request_id": truncateOrNilPtr(requestID, 100),
	}
	if err := r.db.Insert("visit_logs", payload); err != nil {
		slog.Warn("failed to insert visit log", "error", err)
	}
}

// ChatLogParams holds all fields for a chat log insert.
type ChatLogParams struct {
	IP           string
	RequestID    string
	Question     string
	History      interface{}
	ResponseText *string
	Sources      interface{}
	QueryPlan    interface{}
	Model        string
	Status       string
	ErrorMessage *string
}

// InsertChatLog inserts a chat log row. Best-effort — errors are only logged.
func (r *LogsRepository) InsertChatLog(params ChatLogParams) {
	status := params.Status
	if status == "" {
		status = "unknown"
	}
	payload := map[string]interface{}{
		"ip":            truncate(normalizeIP(params.IP), 100),
		"request_id":    truncateOrNil(params.RequestID, 100),
		"question":      truncate(params.Question, 20000),
		"history":       params.History,
		"response_text": params.ResponseText,
		"sources":       params.Sources,
		"query_plan":    params.QueryPlan,
		"model":         truncateOrNil(params.Model, 100),
		"status":        truncate(status, 50),
		"error_message": params.ErrorMessage,
	}
	if err := r.db.Insert("chat_logs", payload); err != nil {
		slog.Warn("failed to insert chat log", "error", err)
	}
}
